### PR TITLE
new camera management

### PR DESCRIPTION
**Pull request by Aneoshun**
See #21 

This PR adds a different way to manage pbuffer used by the cameras. 

**This is still WIP.** 

It requires OSG>=3.6 (to have osgViewer::ViewerBase::setUseConfigureAffinity(bool flag) function), I will probably put this into a `#ifdef` to make it optional. 

Opinions, comments, and ideas are welcome.

### CODE REVIEW

**Comment by costashatz** [comment]
@Aneoshun thanks for this.. I will give it a detailed look most probably next week..

Can you resolve the conflicts in the meanwhile?

**Comment by Aneoshun** [comment]
Sure, I will do.

**Comment by Aneoshun** [comment]
Hi @costashatz, 
Any thought on this? (conflicts are resolved)

File: src/robot_dart/graphics/camera_osr.hpp [hunk 1]
```diff
@@ -46,35 +49,22 @@ namespace robot_dart {
...
-                traits->windowDecoration = false;
-                traits->pbuffer = true;
-                traits->doubleBuffer = true;
-                traits->sharedContext = 0;
-                traits->setUndefinedScreenDetailsToDefaultScreen();
-
-                ::osg::ref_ptr<::osg::GraphicsContext> pbuffer = ::osg::GraphicsContext::createGraphicsContext(traits.get());
-
-                if (pbuffer.valid()) {
+                _osg_viewer->setUseConfigureAffinity(false);
```
Comment (costashatz) [inline src/robot_dart/graphics/camera_osr.hpp]: Didn't you say that this is available only for OSG>=3.6?

File: src/robot_dart/graphics/camera_osr.hpp [hunk 2]
```diff
@@ -27,14 +29,15 @@ namespace robot_dart {
                 width = vp->width();
                 height = vp->height();
 
-                _viewer->image()->readPixels(x, y, width, height, GL_RGB, GL_UNSIGNED_BYTE);
+                _viewer->image()->readPixels(x, y, width, height, GL_LUMINANCE, GL_UNSIGNED_BYTE);
```
Comment (costashatz) [inline src/robot_dart/graphics/camera_osr.hpp]: Why did you change this to `GL_LUMINANCE`? In my computer, I am getting very sharp images, like this one:

cam2

File: src/robot_dart/graphics/camera_osr.hpp [hunk 3]
```diff
@@ -46,35 +49,22 @@ namespace robot_dart {
         class CameraOSR : public BaseGraphics {
 
         public:
-            CameraOSR(const dart::simulation::WorldPtr& world, unsigned int width = 640, unsigned int height = 480, bool shadowed = true) : _world(world), _width(width), _height(height), _frame_counter(0), _enabled(true), _buffer_valid(true), _recording(false), _filename("camera_record.png"), _image(new osg::Image)
+            CameraOSR(const dart::simulation::WorldPtr& world, unsigned int width = 640, unsigned int height = 480, bool shadowed = true) : _world(world), _width(width), _height(height), _frame_counter(0), _enabled(true), _buffer_valid(true), _recording(false), _filename("camera_record.png"), _image(new osg::Image), _shot_done(false)
 
             {
+                if (!global::pbufferManager) {
+                    std::cout << "Error: pbufferManager seems not to be started" << std::endl;
+                    assert(0);
```
Comment (costashatz) [inline src/robot_dart/graphics/camera_osr.hpp]: Maybe you can use the robot_dart assertions for this (and the rest assert(0))..

**Comment by costashatz** [comment]
Hey @Aneoshun ! Thanks again for this!

> Any thought on this? (conflicts are resolved)

I merged with master, formatted and made a few comments. I am still trying to see if there is a better way to do this rather than a global namespace. I will get back to you as soon as possible..

File: src/robot_dart/graphics/camera_osr.hpp (hunk 1 above)
Comment (Aneoshun) [reply src/robot_dart/graphics/camera_osr.hpp]: Yes, it is and it is useful to assign the rendering threads properly. However, it should still run without this command. Maybe we can know the version of OSG installed and put this line in a conditional preprocessor command

File: src/robot_dart/graphics/camera_osr.hpp (hunk 2 above)
Comment (Aneoshun) [reply src/robot_dart/graphics/camera_osr.hpp]: I will revert this, it was just easier to do some tests with it as it forms an image with a single channel (the red one)

File: src/robot_dart/graphics/camera_osr.hpp (hunk 3 above)
Comment (Aneoshun) [reply src/robot_dart/graphics/camera_osr.hpp]: ok

File: src/robot_dart/graphics/camera_osr.hpp (hunk 1 above)
Comment (costashatz) [reply src/robot_dart/graphics/camera_osr.hpp]: This was easy! `osg/Version`

**Comment by costashatz** [comment]
@Aneoshun I do not see any other better way and I do not want to spend too much time on this.. I say that we merge this. I made all the changes that we discussed. Please verify that everything compiles and works in your case and then I will merge..

By the way, I am already working on providing a better (and more customizable) graphics integration using the [Magnum](https://magnum.graphics/) library. You can have a look at the [magnum branch](https://github.com/resibots/robot_dart/tree/magnum); although it is still very early! Using Magnum we would not need separate classes and structures for taking off-screen pictures or creating sensors..

**Comment by Aneoshun** [comment]
@costashatz Thanks a lot for that. Sorry I did not have time to do the changes myself. One of my students tested this branch this morning in his own experiment and it works as expected. So I guess that's ready to merge.
